contentgrid-spring-data-rest/src/test/java/com/contentgrid/spring/data/rest/etag/OptimisticLockingTest.java

```diff
@@ -0,0 +1,433 @@
+import org.springframework.test.web.servlet.MockMvc;
+
+@Slf4j
+@SpringBootTest
+@ContextConfiguration(classes = {
+        InvoicingApplication.class,
+})
+@EnableAutoConfiguration(exclude = EventsAutoConfiguration.class)
+@AutoConfigureMockMvc(printOnlyOnFailure = false)
+@WithMockJwt
+public class OptimisticLockingTest {
+
+    static final String INVOICE_NUMBER_1 = "I-2022-0001";
+    static final String ORG_XENIT_VAT = "BE0887582365";
+    static UUID INVOICE_1_ID;
+    static UUID XENIT_ID;
+
+    static int INVOICE_1_VERSION;
+    static int XENIT_VERSION;
+    static final String INVALID_VERSION = "\"INVALID\"";
+
+    private static final String EXT_ASCII_TEXT = "L'éducation doit être gratuite.";
+    private static final String UNICODE_TEXT = "Some unicode text 💩";
+    private static final String MIMETYPE_PLAINTEXT_LATIN1 = "text/plain;charset=ISO-8859-1";
+    private static final String MIMETYPE_PLAINTEXT_UTF8 = "text/plain;charset=UTF-8";
+
+    @Autowired
+    private MockMvc mockMvc;
+
+    @Autowired
+    CustomerRepository customers;
+
+    @Autowired
+    InvoiceRepository invoices;
+
+    @Autowired
+    InvoiceContentStore invoicesContent;
+
+    @Autowired
+    CustomerContentStore customersContent;
+
+    @BeforeEach
+    void setupTestData() {
+        var xenit = customers.save(
+                new Customer(null, 0, new AuditMetadata(), "XeniT", ORG_XENIT_VAT, null, null, null, new HashSet<>(),
+                        new HashSet<>()));
+
+        XENIT_ID = xenit.getId();
+        XENIT_VERSION = xenit.getVersion();
+
+        var invoice = invoices.save(
+                new Invoice(INVOICE_NUMBER_1, true, false, xenit, new HashSet<>()));
+
+        INVOICE_1_ID = invoice.getId();
+        INVOICE_1_VERSION = invoice.getVersion();
+    }
+
+    void setupContentProperties() {
+        var stream = new ByteArrayInputStream(EXT_ASCII_TEXT.getBytes(StandardCharsets.ISO_8859_1));
+
+        var customer = customers.findById(XENIT_ID).orElseThrow();
+        customersContent.setContent(customer, PropertyPath.from("content"), stream);
+        customer.getContent().setMimetype(MIMETYPE_PLAINTEXT_LATIN1);
+        customer.getContent().setFilename("content.txt");
+        customer = customers.save(customer);
+
+        XENIT_VERSION = customer.getVersion();
+
+        var invoice = invoices.findById(INVOICE_1_ID).orElseThrow();
+        invoicesContent.setContent(invoice, PropertyPath.from("content"), stream);
+        invoice.setContentMimetype(MIMETYPE_PLAINTEXT_LATIN1);
+        invoice.setContentFilename("content.txt");
+        invoice = invoices.save(invoice);
+
+        INVOICE_1_VERSION = invoice.getVersion();
+    }
+
+    void checkETagExists(String url) throws Exception {
+        mockMvc.perform(get(url))
+                .andExpect(status().isOk())
+                .andExpect(headers().etag().exists());
+    }
+
+    void checkETagUnchanged(String url, int original) throws Exception {
```

Comment:
I would also change the type in these arguments:
```suggestion
    void checkETagUnchanged(String url, String etag) throws Exception {
```